data_safe_haven/infrastructure/project_manager.py

```
@@ -105,10 +107,14 @@ def project_settings(self) -> automation.ProjectSettings:
     def stack_settings(self) -> automation.StackSettings:
         return automation.StackSettings(
             config=self.pulumi_project.stack_config,
-            encrypted_key=self.pulumi_project.encrypted_key,
+            encrypted_key=self.pulumi_config.encrypted_key,
             secrets_provider=self.context.pulumi_secrets_provider_url,
         )
 
+    @property
+    def pulumi_project(self) -> DSHPulumiProject:
+        return self.pulumi_config[self.pulumi_project_name]
+
```

Would it be better to do the `create_if_new` here, so we're effectively lazy-loading the project?

---

The trouble with that is, it is only when deploying that we would want to create a `DSHPulumiProject`. In all other cases we expect it to exist and should raise an exception if it doesn't.

---

OK, but are we relying on a default KeyError exception if it's not in the dictionary? Maybe having a `get_project` method or similar on DSHPulumiConfig would help us to raise a more understandable exception?

---

I was thinking over lunch. I think I can get this and the `create_or_select` both working together in a better way.

---

Yes, I think this can all live in one function

---

Fixed in 2d186463